runtime/lib/capi/RuntimeCAPI.cpp

```diff
@@ -1036,14 +1035,34 @@ int64_t __catalyst__rt__array_get_size_1d(QirArray *ptr)
     return qubit_vector_ptr->size();
 }
 
+static bool isDeviceAutomaticQubitManageMode()
+{
+    return RTD_PTR->getDeviceAutoQubitManagementMode();
+}
+
 int8_t *__catalyst__rt__array_get_element_ptr_1d(QirArray *ptr, int64_t idx)
 {
     std::vector<QubitIdType> *qubit_vector_ptr = reinterpret_cast<std::vector<QubitIdType> *>(ptr);
 
     RT_ASSERT(idx >= 0);
     std::string error_msg = "The qubit register does not contain the requested wire: ";
     error_msg += std::to_string(idx);
-    RT_FAIL_IF(static_cast<size_t>(idx) >= qubit_vector_ptr->size(), error_msg.c_str());
+
+    if (static_cast<size_t>(idx) >= qubit_vector_ptr->size()) {
+        if (!isDeviceAutomaticQubitManageMode()) {
+            RT_FAIL(error_msg.c_str());
+        }
+        else {
+            // allocate a new qubit if we are in automatic qubit allocation mode
+            // `idx` is the new user wire index from frontend pennylane
+            // number of currently allocated qubits is `qubit_vector_ptr->size()`
+            while (qubit_vector_ptr->size() <= static_cast<size_t>(idx)) {
+                auto new_alloced_id =
+                    reinterpret_cast<QubitIdType>(__catalyst__rt__qubit_allocate());
+                qubit_vector_ptr->push_back(new_alloced_id);
+            }
+        }
```

Comment:
Other minor suggestion, you could outline this piece into a function to visually distinguish it from the extract element code, as one might not expect to find allocation code inside of it.